packages/core/js-client/src/api.ts

```diff
@@ -74,59 +75,57 @@ export const v5_callFunction = async (
     );
   }
 
-  const argNames = Object.keys(
-    def.arrow.domain.tag === "nil" ? [] : def.arrow.domain.fields,
-  );
+  const schemaFunctionArgs: Record<string, FunctionArg> =
+    def.arrow.domain.tag === "nil" ? {} : def.arrow.domain.fields;
 
-  const schemaArgCount = argNames.length;
+  const schemaFunctionArgEntries = Object.entries(schemaFunctionArgs);
 
-  type FunctionArg = SimpleTypes | ArrowWithoutCallbacks;
+  const schemaArgCount = schemaFunctionArgEntries.length;
 
-  const schemaFunctionArgs: Record<string, FunctionArg> =
-    def.arrow.domain.tag === "nil" ? {} : def.arrow.domain.fields;
+  type FunctionArg = SimpleTypes | ArrowWithoutCallbacks;
 
   // if there are more args than expected in schema (schemaArgCount) then last arg is config
   const config = schemaArgCount < rest.length ? rest.pop() : undefined;
 
   validateAquaConfig(config);
 
   const callArgs = Object.fromEntries<JSONValue | ServiceImpl[string]>(
-    rest.slice(0, schemaArgCount).map((arg, i) => {
-      const argName = argNames[i];
-      const argSchema = schemaFunctionArgs[argName];
+    zip(rest.slice(0, schemaArgCount), schemaFunctionArgEntries).map(
+      ([arg, [argName, argType]]) => {
+        if (argType.tag === "arrow") {
+          if (typeof arg !== "function") {
+            throw new SchemaValidationError(
+              [argName],
+              argType,
+              "function",
+              arg,
+            );
+          }
+
+          return [argName, wrapJsFunction(arg, argType)];
+        }
 
-      if (argSchema.tag === "arrow") {
-        if (typeof arg !== "function") {
+        if (typeof arg === "function") {
           throw new SchemaValidationError(
             [argName],
-            argSchema,
-            "function",
+            argType,
+            "non-function value",
             arg,
           );
         }
 
-        return [argName, wrapJsFunction(arg, argSchema)];
-      }
-
-      if (typeof arg === "function") {
-        throw new SchemaValidationError(
-          [argName],
-          argSchema,
-          "non-function value",
-          arg,
-        );
-      }
-
-      return [argName, js2aqua(arg, argSchema, { path: [def.functionName] })];
-    }),
+        return [argName, js2aqua(arg, argType, { path: [def.functionName] })];
+      },
+    ),
   );
 
   const returnTypeVoid =
     def.arrow.codomain.tag === "nil" || def.arrow.codomain.items.length === 0;
 
   const returnSchema =
     def.arrow.codomain.tag === "unlabeledProduct" &&
-    def.arrow.codomain.items.length === 1
+    def.arrow.codomain.items.length === 1 &&
+    "0" in def.arrow.codomain.items
       ? def.arrow.codomain.items[0]
       : def.arrow.codomain;
```

Comment:
Don't you want to extract this pattern to some function? It is repeated at least 3 times

Reply:
probably possible with this
```
function isArrayWithOneElement<T>(arr: Array<T>): arr is [T] {
  return arr.length === 1 && arr[0] !== undefined;
}
```

Reply:
Will keep it in mind. This PR already bloated too much, i will to this in the next PR.